README
Data-SimplePassword

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	export RUN_HEAVY_TEST=1    # optional
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Data::SimplePassword

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Data-SimplePassword

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Data-SimplePassword

    CPAN Ratings
        http://cpanratings.perl.org/d/Data-SimplePassword

    Search CPAN
        http://search.cpan.org/dist/Data-SimplePassword


COPYRIGHT AND LICENCE

Copyright (C) 2006-2008 Ryo Okamoto

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.



Changes
Revision history for Perl extension Data::SimplePassword.

0.11    Sat Feb  6 07:36:14 UTC 2016

    * removed ambiguous chars from rndpassword script.

0.10    Tue Dec  3 07:10:44 UTC 2013

    * switched to Moo.

0.08    Tue Jun 25 03:03:11 UTC 2013

    * added is_available_provider() method.
    * added new type 'alpha' to rndpassword.
    * improved some tests to reduce test time.
    * added zsh completion sample file. (see extra/)
    * switched to Minilla.

0.07    Wed Feb 23 18:58:18 JST 2011

    * added -u switch to rndpassword(1).
    * optimized for tests.
    * fixed some typos.

0.06    Tue Aug 31 15:37:15 JST 2010

    * added a parameter 'provider' passed to Crypt::Random.
      now you can avoid annoying device block to set the value for example
      to 'udevrandom' (means /dev/urandom), 'rand' etc.

0.05    Tue Oct 13 11:04:46 JST 2009

    * added a few new types for rndpassword(1).
    * added 'endless mode' for rndpassword(1).
    * cleaned code & pod up.

0.04    Sun Jun 15 16:22:32 JST 2008

    * fixed a bug caused memory greediness with too long strings :<
    * improved internal code for PAUSE.

0.03    Sat Jun 14 19:17:30 JST 2008

    * added support for Math::Random::MT::Perl.
    * switched to Module::Build.
    * cleaned up test scripts.
    * added 'binary' option to rndpassword.

0.02    Fri Oct 20 08:57:08 JST 2006

    * added rndpassword command-line tool.

0.01    Fri Aug  4 15:46:42 JST 2006

    * first release.


Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Data-SimplePassword',
    module_name     => 'Data::SimplePassword',
    allow_pure_perl => 0,

    script_files => [glob('script/*'), glob('bin/*')],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.3607
use ExtUtils::MakeMaker;
WriteMakefile
(
          'NAME' => 'Data::SimplePassword',
          'VERSION_FROM' => 'lib/Data/SimplePassword.pm',
          'PREREQ_PM' => {
                           'CLASS' => 0,
                           'Class::Accessor' => 0,
                           'Class::Data::Inheritable' => 0,
                           'Crypt::Random' => 0,
                           'List::MoreUtils' => 0,
                           'Math::Random::MT::Perl' => 0,
                           'Test::More' => '0.88',
                           'UNIVERSAL::require' => 0
                         },
          'INSTALLDIRS' => 'site',
          'EXE_FILES' => [
                           'rndpassword'
                         ],
          'PL_FILES' => {}
        )
;


lib/Data/SimplePassword.pm
#

package Data::SimplePassword;

use Moo;
use MooX::ClassAttribute;

use Carp;
use UNIVERSAL::require;
use Crypt::Random ();

# ABSTRACT: Simple random password generator

our $VERSION = '0.10';

class_has 'class' => (
    is => 'rw',
    default => sub {
	Math::Random::MT->use
	    ? "Math::Random::MT"
	    : Math::Random::MT::Perl->use
		? "Math::Random::MT::Perl"
		: "Data::SimplePassword::exception";
    },
);

has 'seed_num' => (
    is => 'rw',
    default => 1,    # now internal use only, up to 624
);

has 'provider' => (
    is => 'rw',
    trigger => sub {
	my $self = shift;
	my ($provider) = @_;

	$self->is_available_provider( $provider )
	    or croak "RNG provider '$_[0]' is not available on this machine.";
    },
);

has '_default_chars' => (
    is => 'ro',
    default => sub { [ 0..9, 'a'..'z', 'A'..'Z' ] },
);

sub chars {
    my $self = shift;

    if( scalar @_ > 0 ){
	croak "each chars must be a letter or an integer."
	    if scalar grep { length( $_ ) != 1 } @_;

	$self->{_chars} = [ @_ ];
    }

    return wantarray ? @{ $self->{_chars} } : $self->{_chars};
}

sub is_available_provider {
    my $self = shift;
    my ($provider) = @_;

    if( defined $provider and $provider ne '' ){
	my $pkg = sprintf "Crypt::Random::Provider::%s", $provider;
	return eval "use $pkg; $pkg->available()";
    }

    return;
}

sub make_password {
    my $self = shift;
    my $len = shift || 8;

    croak "length must be an integer."
	unless $len =~ /^\d+$/o;

    my @chars = defined $self->chars && ref $self->chars eq 'ARRAY'
	? @{ $self->chars }
	: @{ $self->_default_chars };

    my $gen = $self->class->new(
	map { Crypt::Random::makerandom( Size => 32, Strength => 1, Provider => $self->provider ) } 1 .. $self->seed_num
    );

    my $password;
    while( $len-- ){
	$password .= $chars[ $gen->rand( scalar @chars ) ];
    }

    return $password;
}

{    package    # hide from PAUSE
	Data::SimplePassword::exception;

    use strict;
    use Carp;

    AUTOLOAD { croak "couldn't find any suitable MT classes." }
}

1;

__END__

=encoding utf-8

=head1 NAME

Data::SimplePassword - Simple random password generator

=head1 SYNOPSIS

 use Data::SimplePassword;

 my $sp = Data::SimplePassword->new;
 $sp->chars( 0..9, 'a'..'z', 'A'..'Z' );    # optional

 my $password = $sp->make_password( 8 );    # length

=head1 DESCRIPTION

YA very easy-to-use but a bit strong random password generator.

=head1 METHODS

=over 4

=item B<new>

 my $sp = Data::SimplePassword->new;

Makes a Data::SimplePassword object.

=item B<chars>

 $sp->chars( 0..9, 'a'..'z', 'A'..'Z' );    # default
 $sp->chars( 0..9, 'a'..'z', 'A'..'Z', qw(+ /) );    # b64-like
 $sp->chars( 0..9 );
 my @c = $sp->chars;    # returns the current values

Sets an array of characters you want to use as your password string.

=item B<make_password>

 my $password = $sp->make_password( 8 );    # default
 my $password = $sp->make_password( 1024 );

Makes password string and just returns it. You can set the byte length as an integer.

=back

=head1 EXTRA METHODS

=over 4

=item B<provider>

 $sp->provider("devurandom");    # optional

Sets a type of random number generator, see Crypt::Random::Provider::* for details.

=item B<is_available_provider>

 $sp->is_available_provider("devurandom");

Returns true when the type is available.

=item B<seed_num>

  $sp->seed_num( 32 );    # up to 624

Sets initial seed number (internal use only).

=back

=head1 COMMAND-LINE TOOL

A useful command named rndpassword(1) will be also installed. Type B<man rndpassword> for details.

=head1 DEPENDENCY

Moo, UNIVERSAL::require, Crypt::Random, Math::Random::MT (or Math::Random::MT::Perl),

=head1 SEE ALSO

Crypt::GeneratePassword, Crypt::RandPasswd, String::MkPasswd, Data::Random::String

http://en.wikipedia.org/wiki//dev/random

=head1 REPOSITORY

https://github.com/ryochin/p5-data-simplepassword

=head1 AUTHOR

Ryo Okamoto E<lt>ryo@aquahill.netE<gt>

=head1 COPYRIGHT & LICENSE

Copyright (c) Ryo Okamoto, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
